lib/renderer/src/renderer/core/sokol/SokolTexture.cpp

```diff
@@ -0,0 +1,391 @@
+//
+// Created by nikita on 05.12.2021.
+//
+
+#include <cstring>
+#include <cassert>
+#include <string>
+
+#include "../../lib/sokol_gfx.h"
+#include "../../lib/HandmadeMath.h"
+
+#include "SokolTexture.h"
+#include "../../exception.h"
+
+#define CODE(...) #__VA_ARGS__
+using namespace renderer::core;
+
+sg_shader get_shader_rgba();
+
+typedef struct {
+	hmm_mat4 transform;
+	hmm_vec4 uv_transform;
+} vs_params_t;
+
+typedef struct {
+	hmm_vec4 color;
+} fs_params_t;
+
+
+struct renderer::core::RenderContext {
+	sg_pipeline pip;
+	sg_bindings bind;
+	sg_buffer vertex_buffer;
+	sg_buffer index_buffer;
+	sg_shader shader;
+	sg_image image;
+};
+
+
+SokolTexture::SokolTexture(TextureType texture_type, BlendMode blend_mode, int32_t width, int32_t height,
+						   std::unique_ptr<RenderContext> &&render_context)
+	: _texture_type(texture_type)
+	, _blend_mode(blend_mode)
+	, _width(width)
+	, _height(height)
+	, _color({255, 255, 255, 255})
+	, _render_context(std::move(render_context))
+	, _data(nullptr)
+{}
+
+SokolTexture::~SokolTexture() {
+	delete[] _data;
+}
+
+std::unique_ptr<SokolTexture>
+SokolTexture::create(TextureType texture_type, BlendMode blend_mode, int32_t width, int32_t height) {
+	std::unique_ptr<RenderContext> render_context = std::make_unique<RenderContext>(RenderContext{});
+
+	float vertices[] = {
+		-1.0f,  1.0f, 0.0f, 0.0f,

```

Comment:
nit: we shouldn't really need this vertex buffer, the values can be produced from within the shader based on `gl_VertexID`